### PR TITLE
python/iot3: implement TLS and Websockets

### PR DESCRIPTION
**Pull request by ymorin-orange**
---
**Features**;

* IoT3 Core SDK:
    * Implement MQTT over TLS
    * Implement MQTT over websockets

Closes: #138

---
**How to test**

1. start an environment with python 3.11 in the IoT3 Core SDK directory:
    ```sh
    $ cd python/iot3/
    $ docker container run \
        --rm \
        -ti \
        --network host \
        -e http_proxy \
        -e https_proxy \
        -e no_proxy \
        --user $(id -u):$(id -u) \
        --mount type=bind,source=$(pwd),destination=$(pwd) \
        --workdir $(pwd) \
        --entrypoint /bin/bash \
        python:3.11.9-slim-bookworm
    ```
2. create a /venv/ in a writable location, and activate it:
    ```sh
    (docker) python3 -m venv /tmp/iot3
    (docker) . /tmp/iot3/bin/activate
    ```
3. install the IoT3 Core SDK (notice the trailing dot `.`):
    ```sh
    (docker) pip3 --disable-pip-version-check install .
    ```
4. Start an MQTT client:
    ```sh
    mosquitto_sub \
        -h test.mosquitto.org -p 1883 \
        -t 'test/+/iot3/#' -t 'test/+/iot3/#' \
        -F '%t: %j'
    ```
5. Run the MQTT-only test:
    ```sh
    (docker) ./tests/test-iot3-core-mqtt
    ```

---
**Expected results:**

1. the docker container runs
2. the python /venv/ is created and activated
3. the IoT3 Core SDK is installed
4. 
5. the MQTT messages are reported (the random 16-hexdigit hash
   is the topic where the message was posted):
    ```
    clear: 65c8f6a3a004823e: clear message
    tls: 65c8f6a3a004823e: 3ИСЯЧРТЕD ПЕЅЅД̈GЕ
    ws: 65c8f6a3a004823e: clear message
    wss: 65c8f6a3a004823e: 3ИСЯЧРТЕD ПЕЅЅД̈GЕ
    ```
    * _clear_: native MQTT protocol in clear (no TLS)
    * _tls_: native MQTT protocol over TLS
    * _ws_: MQTT over WebSockets in clear (no TLS)
    * _wss_: MQTT over WebSockets over TLS

### CODE REVIEW

File: python/iot3/src/iot3/core/mqtt.py [hunk 1]
```diff
@@ -7,9 +7,10 @@
 import paho.mqtt.enums
 import paho.mqtt.packettypes
 import paho.mqtt.properties
+import ssl
 import threading
 import time
-from typing import Any, Callable, Optional, TypeAlias, Unpack
+from typing import Any, Callable, Optional, TypeAlias, Union, Unpack
```
Comment (tigroo) [inline python/iot3/src/iot3/core/mqtt.py:13]: useless, no?

File: python/iot3/src/iot3/core/mqtt.py [hunk 2]
```diff
@@ -46,6 +52,48 @@ def __init__(
                             msg_cb function.
         :param span_ctx_cb: The function to obtain a context manager for an
                             OpenTelemetry span.
+
+        There are two ways to connect to the MQTT broker:
+         * via TCP: both host and port must be specified;
+         * via websockets: all of host, port, and websocket_path must
```
Comment (tigroo) [inline python/iot3/src/iot3/core/mqtt.py:58]: `all of the host, ` ?

File: python/iot3/tests/test-iot3-core-all [hunk 3]
```diff
@@ -2,10 +2,14 @@
 
 import iot3.core.mqtt
 import iot3.core.otel
+import random
 import time
 
 def recv(data, topic, payload):
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core-all:8]: PEP 8 2 blank lines possible?

File: python/iot3/tests/test-iot3-core-all [hunk 4]
```diff
@@ -2,10 +2,14 @@
...
-    print(f"{topic[:16]}: {payload[:16]}")
+    print(f"{data}: {topic[:16]}: {payload[:16]}")
+
+client_id = random.randbytes(16).hex()
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core-all:11]: PEP 8 2 blank lines possible?

File: python/iot3/tests/test-iot3-core-mqtt [hunk 5]
```diff
@@ -1,24 +1,82 @@
...
 import iot3.core.mqtt
+import random
 import time
 
 
 def recv(data, topic, payload):
-    print(f"{topic[:16]}: {payload[:16]}")
+    print(f"{data}: {topic[:16]}: {payload.decode()[:32]}")
 
+client_id = random.randbytes(16).hex()
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core-mqtt:11]: PEP 8 2 blank lines possible?

File: python/iot3/tests/test-iot3-core [hunk 6]
```diff
@@ -1,24 +1,34 @@
 #!/usr/bin/env python3
 
 import iot3.core
+import random
 import time
 
 def recv(data, topic, payload):
     print(f"{topic[:16]}: {payload[:16]}")
 
+config = iot3.core.sample_config
```
Comment (tigroo) [inline python/iot3/tests/test-iot3-core:10]: PEP 8 2 blank lines possible?

**Review by tigroo — APPROVED** [review]
test ok: 
```
(iot3) I have no name!@yd-89lzhs3:/home/kvmk8371/Workspace/its-client/python/iot3$ ./tests/test-iot3-core-mqtt
clear: test/4cf3acf8754: clear message
tls: test/4cf3acf8754: 3ИСЯЧРТЕD ПЕЅЅД̈GЕ
ws: test/4cf3acf8754: clear message
wss: test/4cf3acf8754: 3ИСЯЧРТЕD ПЕЅЅД̈GЕ
(iot3) I have no name!@yd-89lzhs3:/home/kvmk8371/Workspace/its-client/python/iot3$ 

```

But a lot of packages out of the iot3 one is impacted by the new development...without clear tests : do we plan special non regression test to be self confident?

File: python/iot3/src/iot3/core/mqtt.py (hunk 2 above)
Comment (ymorin-orange) [reply python/iot3/src/iot3/core/mqtt.py:58]: Not really, no. It must be read as: _all the items in the following list: host, port, and websocket_path_.

File: python/iot3/src/iot3/core/mqtt.py (hunk 1 above)
Comment (ymorin-orange) [reply python/iot3/src/iot3/core/mqtt.py:13]: Indeed, that's a remnant of an interim change that was eventually dropped.

**Comment by ymorin-orange** [comment]
> But a lot of packages out of the iot3 one is impacted by the new development...without clear tests : do we plan special non regression test to be self confident?

That's true that there are some changes in the other components, and that ideally we should test them. However, it is going to be quite a huge investment, and the changes are quite self-contained and relatively obvious. I believe that, if the review did not spot obvious issues, we defer to the future IRL test (next week) to detect any residual bug and open issues in that case. YOLO!

**Review by tigroo — APPROVED** [review]
okay on all the new elements

**Comment by tigroo** [comment]
code read after merge, sorry for the delay :(

**Pull request merged**